### PR TITLE
Modernize CI

### PR DESCRIPTION
**Pull request by patszt**
The CI jobs should be smarter about deciding when to run (eg. change in plugins/ will only trigger plugins rebuild with a downloaded vast package)
Plugins will also be built under macOS clang
Fixed some bugs (cache working incorrectly for multiple rebuilds, wrong CMAKE flags)

### :memo: Reviewer Checklist

Review this pull request by ensuring the following items:

- [ ] All user-facing changes have changelog entries
- [ ] User-facing changes are reflected on [vast.io](https://github.com/tenzir/vast/tree/master/web)

### CODE REVIEW

**Comment by dominiklohmann** [comment]
Realistically I won't get to review this until next week considering the size of the PR, but I'm really looking forward to it. This is great work, @patszt!

File: .github/workflows/vast.yaml [hunk 1]
```diff
@@ -93,73 +103,139 @@ jobs:
...
+            run_vast=true
+          elif ! is_unchanged tools/; then
+            run_vast=true
+          elif ! is_unchanged vast/; then
+            run_vast=true
+          elif ! is_unchanged vast.yaml.example; then
+            run_vast=true
+          elif ! is_unchanged version.json; then
+            run_vast=true
+          fi
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:173]: For all instances of is_unchanged here, do you think it'd make sense to group them together, i.e., `if ! is_unchanged cmake/ CMakeLists.txt libvast/ ...`?

File: .github/workflows/vast.yaml [hunk 2]
```diff
@@ -188,626 +264,78 @@ jobs:
...
+  vast-io:
     needs:
-      - determine-version
-    if: github.event_name != 'workflow_dispatch'
-    name: Debian ${{ matrix.configure.tag }} (${{ matrix.build.compiler }})
+      configure
+      changelog
+      docker
+    if: ${{ needs.configure.outputs.run-vast-io == 'true' }}
+    name: vast-io
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:273]:
```suggestion
    name: vast.io
```

File: .github/workflows/vast.yaml [hunk 3]
```diff
@@ -1079,23 +607,448 @@ jobs:
...
+    needs:
+      - vast
+      - configure
+    if: ${{ always() && needs.configure.outputs.run-vast-plugins == 'true' }}
+    runs-on: ${{ matrix.plugin.os }}
+    container: ${{ matrix.plugin.container }}
+    strategy:
+      fail-fast: false
+      matrix:
+        plugin:
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:835]: If you add a second record on this level you can make the matrix two-dimensional. I'm not sure whether that makes much sense, but it is an option here.

File: .github/workflows/vast.yaml [hunk 4]
```diff
@@ -1079,23 +607,448 @@ jobs:
...
+            target: parquet
+            path: plugins/parquet
+            os: ubuntu-20.04
+            container: debian:bullseye-slim
+          - name: Parquet macOS
+            target: parquet
+            path: plugins/parquet
+            os: macos-latest
+            container:
+          - name: Web Debian
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:887]: If I'm not mistaken then the PCAP plugin is missing from this list.

File: .github/workflows/vast.yaml [hunk 5]
```diff
@@ -1079,23 +607,448 @@ jobs:
...
+      CMAKE_GENERATOR: Ninja
+      CMAKE_MAKE_PROGRAM: ninja
+      CTEST_OUTPUT_ON_FAILURE: YES
+      DEBIAN_FRONTEND: noninteractive
+      HOMEBREW_GITHUB_API_TOKEN: ${{ github.token }}
+      HOMEBREW_NO_ANALYTICS: 1
+      HOMEBREW_NO_INSTALL_CLEANUP: 1
+      HOMEBREW_NO_AUTO_UPDATE: 1
+      DESTDIR: "${{ github.workspace }}"
+    name: ${{ matrix.plugin.name }} Plugin
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:914]: This now renders as `$plugin $os Plugin`, which reads a bit weird. In order to align with the other steps, can we make this `VAST Plugins ($plugin, $os)`?

File: .github/workflows/vast.yaml [hunk 6]
```diff
@@ -1079,23 +607,448 @@ jobs:
...
+          submodules: true
+      - name: Install Dependencies macos
+        if: ${{ matrix.plugin.os == 'macos-latest' }}
+        run: |
+          brew install \
+            ${{ join(matrix.plugin.dependencies, ' ') }} \
+            curl
+          ./scripts/macOS/install-dev-dependencies.sh
+          python3 -m pip install --upgrade pip
+      - name: Install Dependencies ubuntu
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:947]:
```suggestion
      - name: Install Dependencies (Debian)
```

File: .github/workflows/vast.yaml [hunk 7]
```diff
@@ -1079,23 +607,448 @@ jobs:
...
+      - name: Install git for debian
+        if: ${{ matrix.plugin.os == 'ubuntu-20.04' }}
+        run: |
+          apt-get update
+          apt-get -y install git
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          submodules: true
+      - name: Install Dependencies macos
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:939]:
```suggestion
      - name: Install Dependencies (macOS)
```

File: .github/workflows/vast.yaml [hunk 8]
```diff
@@ -1079,23 +607,448 @@ jobs:
...
+            PACKAGE_SUFFIX=Release-GCC
+           elif ${{ matrix.plugin.os == 'macos-latest' }}; then
+            CC=clang
+            CXX=clang++
+            PACKAGE_SUFFIX=Release-Clang
+           fi
+           echo "CC=${CC}" >> $GITHUB_ENV
+           echo "CXX=${CXX}" >> $GITHUB_ENV
+           echo "PACKAGE_SUFFIX=${PACKAGE_SUFFIX}" >> $GITHUB_ENV
+      - name: Install git for debian
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:930]:
```suggestion
      - name: Install Git (Debian)
```

File: .github/workflows/vast.yaml [hunk 9]
```diff
@@ -1079,23 +607,448 @@ jobs:
...
+      - name: Configure GCS Credentials
+        if: ${{ needs.configure.outputs.run-vast == 'false' }}
+        uses: google-github-actions/setup-gcloud@v0
+        with:
+          service_account_key: ${{ secrets.GCP_SA_KEY }}
+          export_default_credentials: true
+      - name: Download VAST
+        if: ${{ needs.configure.outputs.run-vast == 'false' }}
+        run: |
+          gsutil cp "gs://${{ secrets.GCS_BUCKET }}/${{ env.PACKAGE_NAME }}.tar.gz" .
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:979]: How do we test this? Just by merging and opening another PR that only modifies a plugin?

File: .github/workflows/vast.yaml [hunk 10]
```diff
@@ -930,16 +457,17 @@ jobs:
...
           make docker TARGET=clean
 
-  build-nix:
+  vast-nix:
     needs:
-      - determine-version
-    name: Nix Static (${{ matrix.nix.target }})
+      - configure
+    if: ${{ needs.configure.outputs.run-vast-nix == 'true' }}
+    name: VAST (Nix, ${{ matrix.nix.target }})
```
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:470]: Reminder to check for conflicts with #2684 before merging, and maybe leaving a note on that PR to double-check.

**Review by dominiklohmann — APPROVED** [review]
This is great. I mostly have some minor requests on this. I tried to review this very thoroughly, but I think some of the changes will only be testable after merging.

File: .github/workflows/vast.yaml (hunk 9 above)
Comment (patszt) [reply .github/workflows/vast.yaml:979]: I have created a new branch from this one. Then i have created PR and set the merge target to current PR's branch.
I have changed something in one plugin to trigger changes in /plugins. I have also changed some SHAs but i am not sure if it was to affect this job. It was working under such circumstances

Comment (dominiklohmann) [reply .github/workflows/vast.yaml:979]: Okay, that should be good enough to have some confidence. 👍

File: scripts/debian/install-dev-dependencies.sh [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    python3-dev \
+    python3-pip \
+    python3-venv \
+    robin-map-dev \
+    software-properties-common \
+    wget
+wget "https://apache.jfrog.io/artifactory/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')/apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb"
+apt-get -y --no-install-recommends install ./apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb
+apt-get update
+apt-get -y --no-install-recommends install libarrow-dev=9.0.0-1 libprotobuf-dev libparquet-dev=9.0.0-1
```
Comment (dominiklohmann) [inline scripts/debian/install-dev-dependencies.sh:43]: Note that this silently conflicts with #2685. Depending on which gets merged first we may have to make adaptions to the other.

File: .github/workflows/vast.yaml (hunk 4 above)
Comment (patszt) [reply .github/workflows/vast.yaml:887]: Seems to be missing from debian as well. I guess it should be added for both, rgiht?

Comment (dominiklohmann) [reply .github/workflows/vast.yaml:887]: Yes.

File: .github/workflows/vast.yaml (hunk 3 above)
Comment (patszt) [reply .github/workflows/vast.yaml:835]: Definitely, less repetition. I will try to get this working

File: .github/workflows/vast.yaml (hunk 5 above)
Comment (patszt) [reply .github/workflows/vast.yaml]: done

File: .github/workflows/vast.yaml (hunk 1 above)
Comment (patszt) [reply .github/workflows/vast.yaml:174]: Yes, i think this is better without this many if elifs :laughing:

**Comment by dit7ya** [comment]
I think jobs should run when there is a change in the definition of the jobs file as well. For example, after rebasing #2684 is skipping the Nix build :shrug: and I even manually triggering one single job from this page skips it - https://github.com/tenzir/vast/actions/runs/3434255497

**Pull request merged**